src/pybind/mgr/cephadm/services/nfs.py

```diff
@@ -156,47 +156,12 @@ def create_rados_config_obj(self, clobber=False):
 
     def get_ganesha_conf(self):
         # type: () -> str
-        return '''# generated by cephadm
-NFS_CORE_PARAM {{
-        Enable_NLM = false;
-        Enable_RQUOTA = false;
-        Protocols = 4;
-}}
-
-MDCACHE {{
-        Dir_Chunk = 0;
-        NParts = 1;
-        Cache_Size = 1;
-}}
-
-EXPORT_DEFAULTS {{
-        Attr_Expiration_Time = 0;
-}}
-
-NFSv4 {{
-        Delegations = false;
-        RecoveryBackend = 'rados_cluster';
-        Minor_Versions = 1, 2;
-}}
-
-RADOS_KV {{
-        UserId = "{user}";
-        nodeid = "{nodeid}";
-        pool = "{pool}";
-        namespace = "{namespace}";
-}}
-
-RADOS_URLS {{
-        UserId = "{user}";
-        watch_url = "{url}";
-}}
-
-%url    {url}
-'''.format(user=self.get_rados_user(),
-           nodeid=self.get_daemon_name(),
-           pool=self.spec.pool,
-           namespace=self.spec.namespace if self.spec.namespace else '',
-           url=self.spec.rados_config_location())
+        context = dict(user=self.get_rados_user(),
+                       nodeid=self.get_daemon_name(),
+                       pool=self.spec.pool,
+                       namespace=self.spec.namespace if self.spec.namespace else '',
+                       url=self.spec.rados_config_location())
+        return self.mgr.template.render('services/nfs/ganesha.conf.j2', context)
```

Comment:
hm. This reduces the test coverage here as the variable names are no longer verified by mypy. 

Is this function already executed when running pytest?

Reply:
Yes, if a service has a test then the templating will be performed.
I enable strict undefined detection in Jinja2, so if there is an undefined variable (without default value in the template), then an exception will be raised.

```
  File "/ceph/src/pybind/mgr/cephadm/services/iscsi.py", line 59, in create
    igw_conf = self.mgr.template.render('services/iscsi/iscsi-gateway.cfg.j2', context)
  File "/ceph/src/pybind/mgr/cephadm/template.py", line 65, in render
    return self.engine.render(name, ctx)
  File "/ceph/src/pybind/mgr/cephadm/template.py", line 47, in render
    raise UndefinedError(e.message)
cephadm.template.UndefinedError: 'doc' is undefined
```

This should be able to catch some error before real deploy.